columnar/src/backend/columnar/columnar_cache.c

```diff
@@ -0,0 +1,557 @@
+        {
+          size -= candidate->length;
+
+          /* Remove the entry from candidates. */
+          if (prev)
+          {
+            prev->next = candidate->next;
+          }
+        }
+      }
+    }
+
+    lastCount = nextLowestCount;
+    nextLowestCount = PG_UINT64_MAX;
+  }
+
+  MemoryContextSwitchTo(oldContext);
+  MemoryContextDelete(evictionContext);
+}
+
+/*
+ * ColumnarAddCacheEntry
+ *
+ * Adds a cache entry, or updates an existing entry.
+ */
+void ColumnarAddCacheEntry(uint64 relId, uint64 stripeId, uint64 chunkId, uint32 columnId, void *data)
+{
+  if (columnar_enable_page_cache == false)
+  {
+    return;
+  }
+
+  MemoryContext oldContext = MemoryContextSwitchTo(ColumnarCacheMemoryContext());
+
+  ColumnarCacheEntry *entry = ColumnarFindInCache(relId, stripeId, chunkId, columnId);
+
+  if (entry != NULL)
+  {
+    /* Free up any existing stored data, everything else will be overwritten. */
+    StringInfo str = entry->store;
+    if (str->data)
+    {
+      pfree(str->data);
+    }
+
+    pfree(str);
+
+    totalAllocationLength -= entry->length;
+  }
+  else
+  {
+    entry = palloc0(sizeof(ColumnarCacheEntry));
+
+    entry->relId = relId;
+    entry->stripeId = stripeId;
+    entry->chunkId = chunkId;
+    entry->columnId = columnId;
+    entry->creationTime = entry->lastAccessTime = time(NULL);
+    entry->readCount = 0;
+
+    /* Add the entry into the list if it is not already there. */
+    if (entry->prev == NULL)
+    {
+      /* Take the place as the last entry. */
+      entry->prev = lastEntry;
+      lastEntry = entry;
+
+      if (lastEntry->prev)
+      {
+        lastEntry->prev->next = entry;
+      }
+
+      if (cacheStorage == NULL)
+      {
+        /* This is the first entry in the cache. */
+        cacheStorage = entry;
+      }
+    }
+  }
+
+  uint64 size;
+  entry->store = CopyStringInfo(data, &size);
+  entry->length = size;
+
+  totalAllocationLength += size;
+
+  if (totalAllocationLength >= statistics.maximumCacheSize)
+  {
+    statistics.maximumCacheSize = totalAllocationLength;
+  }
+
+  /* If we are over our cache allocation, clear until we are at 90%. */
+  if (totalAllocationLength >= (columnar_page_cache_size * 1024 * 1024))
+  {
+    EvictCache((columnar_page_cache_size * 1024 * 1024 * .1) + (totalAllocationLength - (columnar_page_cache_size * 1024 * 1024)));
+  }
+
+  statistics.writes++;
+
+  MemoryContextSwitchTo(oldContext);
+}
+
+/*
+ * ColumnarRetrieveCache
+ *
+ * Search for a cache entry, returning NULL if not found.  If found,
+ * make a copy in the current memory context and return it.
+ */
+void *ColumnarRetrieveCache(uint64 relId, uint64 stripeId, uint64 chunkId, uint32 columnId)
+{
+  if (columnar_enable_page_cache == false)
+  {
+    return NULL;
+  }
+
+  ColumnarCacheEntry *entry = ColumnarFindInCache(relId, stripeId, chunkId, columnId);
+
+  if (entry == NULL)
+  {
+    return NULL;
+  }
+
+  void *chunkCopy = CopyStringInfo(entry->store, NULL);
```

Comment:
Can you skip freeing decompressed chunk data in query, and just keep it in cache until it is invalidated ?

Reply:
only if we change `EndChunkGroupRead(ChunkGroupReadState *chunkGroupReadState)` to not deallocate the memory on a read.  I was wary to do this because we would need to track full state of caching throughout the application.  currently, we do not track this state through the full applications.

places that would need to change:

* a check in `DeserializeChunkData` for whether to cache would need to keep the cache status
* a check in `EndChunkGroupRead` against every column as to whether or not to free the memory (since we shouldn't be attempting to read free'd memory)
* `ChunkData` would need to track each column's cache status

otherwise, we will have huge memory bloat, which is one of two problems I was trying to solve.  the second being that this was a _lightweight_ change without making major changes throughout the extension.

If we want to revisit this being _lightweight_ we can, but I'd still rather this be a small change rather than a much bigger effort.

Reply:
OK, will discuss this but i don't think that it should be complicated.